File: packages/kotlin/src/main/kotlin/com/natsuneko/catalyst/models/CatalystContest.kt
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

package com.natsuneko.catalyst.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Lifecycle state of a contest
 */
@Serializable
enum class CatalystContestState {
    @SerialName("draft")
    DRAFT,

    @SerialName("published")
    PUBLISHED,

    @SerialName("opening")
    OPENING,

    @SerialName("closing")
    CLOSING,

    @SerialName("voting")
    VOTING,

    @SerialName("electing")
    ELECTING,

    @SerialName("closed")
    CLOSED
}

/**
 * Voting window and configuration for a contest
 *
 * Note: [since]/[until] are ISO8601 strings (not `Instant`) because the real API may
 * return an empty string when the window is unset, which `Instant` cannot parse.
 */
@Serializable
data class CatalystContestVoting(
    val since: String,
    val until: String,
    val maxVotes: Int,
    val isEnable: Boolean
)

/**
 * A prize rank/tier within a contest
 */
@Serializable
data class CatalystContestRank(
    val id: String,
    val name: String,
    val description: String? = null,
    val count: Int,
    val prize: String,
    val winners: List<EgeriaUser>
)

/**
 * A contest hosted on the platform
 *
 * Note: [winnersOpenAt], [winnersMessageSendAt], [publishedAt], [since], and [until] are
 * ISO8601 strings (not `Instant`) because the real API may return an empty string when
 * these fields are unset, which `Instant` cannot parse.
 */
@Serializable
data class CatalystContest(
    val slug: String,
    val draft: Boolean,
    val state: CatalystContestState,
    val title: String,
    val description: String,
    val theme: String,
    val terms: String,
    val headerUrl: String,
    val bannerUrl: String,
    val organizer: EgeriaUser,
    val winnersOpenAt: String,
    val winnersMessageSendAt: String,
    val publishedAt: String,
    val since: String,
    val until: String,
    val allowSensitive: Boolean,
    val maxMediaPerEntry: Int? = null,
    val voting: CatalystContestVoting,
    val ranks: List<CatalystContestRank>
)

/**
 * The current user's remaining vote allowance for a contest
 */
@Serializable
data class CatalystUserVoteRights(
    val remaining: Int,
    val statuses: List<String>
)

/**
 * Wrapper for the `{"contest": {...}}` response shape
 */
@Serializable
internal data class CatalystContestWrapper(val contest: CatalystContest)

/**
 * Wrapper for the `{"contests": [...]}` response shape
 */
@Serializable
internal data class CatalystContestsWrapper(val contests: List<CatalystContest>)

File: packages/swift/Sources/CatalystSwift/Models/CatalystContest.swift
// Licensed under the MIT License
//
// Copyright (c) 2025 Natsune Mochizuki

import Foundation

public enum CatalystContestState: String, Codable, Sendable, Equatable, Hashable {
  case draft
  case published
  case opening
  case closing
  case voting
  case electing
  case closed
}

public struct CatalystContestVoting: Decodable, Sendable, Equatable, Hashable {
  /// ISO8601 string, or empty when unset.
  public let since: String
  /// ISO8601 string, or empty when unset.
  public let until: String
  public let maxVotes: Int
  public let isEnable: Bool
}

public struct CatalystContestRank: Decodable, Sendable, Equatable, Hashable {
  public let id: String
  public let name: String
  public let description: String?
  public let count: Int
  public let prize: String
  public let winners: [EgeriaUser]
}

public struct CatalystContest: Decodable, Sendable, Equatable, Hashable {
  public let slug: String
  public let draft: Bool
  public let state: CatalystContestState
  public let title: String
  public let description: String
  public let theme: String
  public let terms: String
  public let headerUrl: String
  public let bannerUrl: String
  public let organizer: EgeriaUser
  /// ISO8601 string, or empty when unset.
  public let winnersOpenAt: String
  /// ISO8601 string, or empty when unset.
  public let winnersMessageSendAt: String
  /// ISO8601 string, or empty when unset.
  public let publishedAt: String
  /// ISO8601 string, or empty when unset.
  public let since: String
  /// ISO8601 string, or empty when unset.
  public let until: String
  public let allowSensitive: Bool
  public let maxMediaPerEntry: Int?
  public let voting: CatalystContestVoting
  public let ranks: [CatalystContestRank]
}

/// The current user's remaining vote allowance for a contest.
public struct CatalystUserVoteRights: Decodable, Sendable, Equatable, Hashable {
  public let remaining: Int
  public let statuses: [String]
}

public struct CatalystContestWrapper: Decodable, Sendable {
  public let contest: CatalystContest
}

public struct CatalystContestsWrapper: Decodable, Sendable {
  public let contests: [CatalystContest]
}
